source/production/F0.Analyzers/CodeAnalysis/CodeRefactorings/ObjectInitializer.cs

```diff
@@ -102,18 +109,15 @@ private static async Task<Document> CreateObjectInitializer(Document document, S
 			return documentEditor.GetChangedDocument();
 		}
 
-		private static IEnumerable<ISymbol> GetMutableMembers(ref TypeInfo typeInfo, Compilation compilation)
+		private static IEnumerable<ISymbol> GetMutableMembers(in TypeInfo typeInfo, Compilation compilation)
 		{
 			var members = new HashSet<ISymbol>(SymbolNameComparer.Instance);
 			var mutableMembers = new List<ISymbol>();
 
 			var type = typeInfo.Type;
-			var objectType = compilation.GetSpecialType(SpecialType.System_Object);
-			var valueType = compilation.GetSpecialType(SpecialType.System_ValueType);
-
-			while (type is { } && type != objectType && type != valueType)
+			while (type is not null && type.SpecialType is not SpecialType.System_Object and not SpecialType.System_ValueType)
 			{
-				var instanceMembers = type.GetMembers().Where(s => !s.IsStatic);
+				var instanceMembers = type.GetMembers().Where(s => !s.IsStatic).ToImmutableArray();
```

Comment:
Is this just "nice" or serves it the purpose of safe guarding the result? Would the Enumerator with Linq Where not handle this?

I did not know about this one, so if I am incorrect please elaborate.

Reply:
The intermediate materialization has no logical impact. This was one of the issues reported by R#/Rider: "Possible multiple enumeration of IEnumerable﻿".
We actually do enumerate the `instanceMembers` enumearble twice afterwards, to filter both accessible fields and properties with accessible set-accessors.

And I decided to use `ToImmutableArray` over `ToArray`, since Roslyn heavily uses the _Immutable_ collections and I wanted to stick with that "domain".

Benchmark without intermediate materialization
```
|                  Method |       Runtime |     Mean |     Error |    StdDev |      Min |      Max |   Median |   Gen 0 |  Gen 1 | Gen 2 | Allocated |
|------------------------ |-------------- |---------:|----------:|----------:|---------:|---------:|---------:|--------:|-------:|------:|----------:|
| CreateObjectInitializer |    .NET 4.7.2 | 1.110 ms | 0.0048 ms | 0.0045 ms | 1.101 ms | 1.120 ms | 1.110 ms | 54.6875 | 9.7656 |     - | 339.33 KB |
| CreateObjectInitializer | .NET Core 5.0 | 1.045 ms | 0.0096 ms | 0.0090 ms | 1.033 ms | 1.060 ms | 1.043 ms | 54.6875 | 9.7656 |     - | 339.33 KB |
```
Benchmark with intermediate materialization via `ToImmutableArray`
```
|                  Method |       Runtime |     Mean |     Error |    StdDev |      Min |      Max |   Median |   Gen 0 |  Gen 1 | Gen 2 | Allocated |
|------------------------ |-------------- |---------:|----------:|----------:|---------:|---------:|---------:|--------:|-------:|------:|----------:|
| CreateObjectInitializer |    .NET 4.7.2 | 1.045 ms | 0.0186 ms | 0.0174 ms | 1.026 ms | 1.075 ms | 1.036 ms | 54.6875 | 9.7656 |     - | 340.21 KB |
| CreateObjectInitializer | .NET Core 5.0 | 1.048 ms | 0.0202 ms | 0.0198 ms | 1.015 ms | 1.084 ms | 1.045 ms | 54.6875 | 9.7656 |     - | 340.21 KB |
```
Benchmark source: see https://github.com/Flash0ver/F0.Analyzers/blob/1fe72fd388988d9027dd604dc76dbec648a66053/source/performance/F0.Analyzers.Benchmarks/CodeAnalysis/CodeRefactorings/ObjectInitializerBenchmarks.cs

Reply:
alright did not see the part where it is enumerated, so it is more a style question then, thanks for pointing this out.